src/compile/data/optimize.ts

```diff
@@ -114,6 +115,35 @@ function getLeaves(roots: DataFlowNode[]) {
   return leaves;
 }
 
+/**
+ * ParseNodes are not mergeable only if their `this._parse`s have a common key but different values for that key.
+ */
+export function isParseMergeable(x: ParseNode, y: ParseNode) {
+  const xParse = x.parse;
+  const yParse = y.parse;
+  for (const k of keys(xParse)) {
+    if (k in yParse && yParse[k] !== xParse[k]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+export function mergeParse(node: DataFlowNode) {
+  const parseChildren = node.children.filter((x): x is ParseNode => x instanceof ParseNode);
+
+  if (parseChildren.length > 1) {
```

Comment:
What if there are more than 2 children and only two of them are mergeable.  
Do we abort the merge somehow?

Reply:
I think we should walk over the children and extract all the parse that is consistent and move it down and only leave what cannot be merged.

Reply:
That's better but might be way more complicated.  

If it's too hard, I think it's okay to do the abort version for now. 
FWIW, our scale, axis, and legend logic abort when one of them can't be merged. (If easy, we don't have to do the same here, but I'm just saying it's better than before already.)

Reply:
It's pretty easy. Create a new node and add to it all the parses from the nodes you are looking at. Then remove empty parse nodes.

Reply:
So it looks like the current logic already merges all mergeable children and simply ignores unmergable ones, but it doesn't introduce a new node to apply the shared parse. 

The question is whether this is correct if we don't introduce an intermediate node.  (I think it's probably wrong -- as something will get parse in the parent, which now merged conflicting parse, if we try to parse the field into another format later, then it won't get the original input?) 

If this is wrong, we may have to revise our approach.

Reply:
Yeah, it is wrong to parse the same field again unless you parse a string as a number or a number as a date later.

Reply:
Hmm, then shouldn't we try to do this parse merging in a bottom-up fashion -- while collecting potential conflicting information along the way?

Reply:
If we only use this information during the parse-merge logic, I think we don't have to make it a part of the dataflow. We can actually collect and store them as a part of this parse-merge logic.

Reply:
From slack conversation, sounds like we can pretty much always merge. Please correct this.